src/main/java/dev/nolij/zson/Zson.java

```diff
@@ -230,12 +233,32 @@ public static Map<String, ZsonValue> obj2Map(@Nullable Object object) {
 		Map<String, ZsonValue> map = object();
 		for (Field field : object.getClass().getDeclaredFields()) {
 			if(!shouldInclude(field, true)) continue;
-			ZsonField value = field.getAnnotation(ZsonField.class);
-			String comment = value == null ? null : value.comment();
+			ZsonField annotation = field.getAnnotation(ZsonField.class);
+			String comment = annotation == null ? NO_COMMENT : annotation.comment();
 			try {
 				boolean accessible = field.isAccessible();
 				if (!accessible) field.setAccessible(true);
-				map.put(field.getName(), new ZsonValue("\0".equals(comment) ? null : comment, field.get(object)));
+
+				Object value = field.get(object);
+				if(value == null) {
+					map.put(field.getName(), new ZsonValue(comment, null));
+				} else if(value.getClass().isArray()) {
+					List<Object> list = new ArrayList<>();
+					int length = Array.getLength(value);
+					for (int i = 0; i < length; i++) {
+						list.add(Array.get(value, i));
+					}
+					value = list;
+				} else if(value instanceof Iterable) {
+					List<Object> list = new ArrayList<>();
+					for (Object o : (Iterable<?>) value) {
+						list.add(o);
+					}
+					value = list;
+				} else if(value instanceof Map) {
+					value = obj2Map(value);
+				}
+				map.put(field.getName(), new ZsonValue(comment, value));
```

Comment:
You could remove one of the `map.put` calls and save a couple more bytes. :wink:
Indentation is not correct but this gets my point across.

```java
    if(value != null) {
        if(value is array) value = new thing;
        else if(value instanceof Map) value = obj2Map(value);
    }
    map.put(field.getName(), new ZsonValue(comment, value));
```

Reply:
Good catch, that would've put null value in the map and then immediately overwritten it